src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixEstimator.java

```diff
@@ -0,0 +1,249 @@
+
+		if(op.hashType == HashType.ExpHash && op.operatorType == AUType.COUNT_DISTINCT_ESTIMATE_KMV) {
+			throw new DMLException(
+				"Invalid hashing configuration using " + HashType.ExpHash + " and " + AUType.COUNT_DISTINCT_ESTIMATE_KMV);
+		}
+
+		if(op.operatorType == AUType.COUNT_DISTINCT_ESTIMATE_HYPER_LOG_LOG)
+			throw new NotImplementedException("HyperLogLog not implemented");
+
+		// Just use naive implementation if the size is small.
+		if(in.getNumRows() * in.getNumColumns() < minimumSize) {
+			return CountDistinctValuesNaive(in);
+		}
+
+		switch(op.operatorType) {
+			case COUNT_DISTINCT:
+				return CountDistinctValuesNaive(in);
+			case COUNT_DISTINCT_ESTIMATE_KMV:
+				return CountDistinctValuesKVM(in, op);
+			case COUNT_DISTINCT_ESTIMATE_HYPER_LOG_LOG:
+				return CountDistinctHyperLogLog(in);
+			default:
+				throw new DMLException("Invalid or not implemented Estimator Type");
+		}
+	}
+
+	/**
+	 * Naive implementation of counting Distinct values. Benefit Precise, but Uses memory, on the scale of input.
+	 * 
+	 * @param in  the input matrix to count number distinct values in
+	 * @return the distinct count
+	 */
+	private static int CountDistinctValuesNaive(MatrixBlock in) {
+		// Make a hash set to contain all the distinct.
+		// Memory usage scale linear with number of distinct values.
+
+		Set<Double> distinct = new HashSet<Double>();
+
+		for(int c = 0; c < in.getNumColumns(); c++) {
+			for(int r = 0; r < in.getNumRows(); r++) {
+				distinct.add(in.getValue(r, c));
+			}
+		}
+
+		return distinct.size();
+
+		// debugging
+		// LOG.debug("Distinct actual array: " + Arrays.toString(distinct.toArray()));
+		// LOG.debug("Number destinct: " + out);
+
+	}
+
+	/**
+	 * KMV synopsis(for k minimum values) Distinct-Value Estimation
+	 * 
+	 * Kevin S. Beyer, Peter J. Haas, Berthold Reinwald, Yannis Sismanis, Rainer Gemulla: On synopses for distinct‐value
+	 * estimation under multiset operations. SIGMOD 2007
+	 * 
+	 * @param in
+	 * @return the distinct count
+	 */
+	private static int CountDistinctValuesKVM(MatrixBlock in, EstimatorOperator op) {
+
+		// D is the number of possible distinct values in the MatrixBlock.
+		// As a default we set this number to numRows * numCols
+		int D = in.getNumRows() * in.getNumColumns();
+
+		// To ensure that the likelihood to hash to the same value we need O(D^2) bits
+		// to hash to assign.
+		// This is handled by our custom bit array class.
+		long tmp = (long) D * (long) D;
+		LOG.debug("M not forced to int size: " + tmp);
+		int M = (tmp > (long) Integer.MAX_VALUE) ? Integer.MAX_VALUE : (int) tmp;
+
+		// the estimator is asymptotically unbiased as k becomes large
+		// memory scales with k.
+		// D >> k >> 0
+		int k = 1024;
+
+		PriorityQueue<Integer> smallestHashes = new PriorityQueue<>(k, Collections.reverseOrder());
+
+		// cache for common elements.
+		Set<Integer> cache = new HashSet<>();
```

Comment:
Positioning this structure as a cache is somewhat misleading, as you seem to use it for containment checks to avoid uniqueness in the priority queue. Maybe create a private class that covers both under a simple PriorityQueue that also ensures uniqueness.